librz/il/core_theory_vm.h

```diff
@@ -0,0 +1,143 @@
+#ifndef BUILD_CORE_THEORY_VM_H
+#define BUILD_CORE_THEORY_VM_H
+
+#include "definitions/wrapper.h"
+#include "core_theory_opcodes.h"
+
+// TODO replace with rz hash table
+// TODO find a proper way to handle var and val
+#define VM_MAX_VAR  2048
+#define VM_MAX_VAL  1024
+#define VM_MAX_LAB  1024
+#define VM_MAX_EFF  1024
+#define VM_MAX_FLG  1024
+#define VM_MAX_TEMP 32
+
+typedef struct rz_il_vm_t *RzILVM;
+typedef void (*RzILOpHandler)(RzILVM vm, RzILOp op);
+typedef void (*RzILVmHook)(RzILVM vm, RzILOp op);
+
+// Main structure of VM
+struct rz_il_vm_t {
+	// record the Var and Val in VM
+	// persistent Variable, Value and effects
+	RzILBag vm_global_value_set;
+	RzILVar *vm_global_variable_list;
+
+	// Memory : should support memory switch
+	Mem *mems;
+	int var_count, val_count, mem_count, lab_count;
+	int addr_size;
+	int data_size;
+
+	RzILTemp *temp_value_list;
+
+	// binding relationships
+	HtPP *vm_global_bind_table;
+	HtPP *vm_global_label_table; // [label->name]->label
+
+	// core theory opcodes
+	//      key : Address (BitVector)
+	//      val : opcode struct list (RzList of RzILOp)
+	HtPP *ct_opcodes;
```

Comment:
Caching the opcode lists like this will most likely cause many problems in the long run, in particular whenever the underlying code changes, the respective entries in here would have to be invalidated, which is tricky to do.
For now, I would go for the simplest possible solution and always disassemble and lift from scratch every time a step should be performed.

Reply:
Yeah, that's may be a problem. Which function should I call to disassemble and lift from scratch ? I used`rz_analysis_op(analysis, &op, addr, code, sizeof(code), RZ_ANALYSIS_OP_MASK_ESIL | RZ_ANALYSIS_OP_MASK_HINT);` to do this, is it right ?